apollo-runtime/src/main/java/com/apollographql/apollo/internal/batch/BatchPoller.kt

```diff
@@ -0,0 +1,82 @@
+package com.apollographql.apollo.internal.batch
+
+import com.apollographql.apollo.api.ScalarTypeAdapters
+import com.apollographql.apollo.api.internal.ApolloLogger
+import com.apollographql.apollo.exception.ApolloException
+import com.apollographql.apollo.interceptor.ApolloInterceptor
+import okhttp3.Call
+import okhttp3.HttpUrl
+import java.util.*
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import kotlin.collections.ArrayList
+
+/**
+ * Class responsible for holding the queue of requests, and execute them as a batch every X ms,
+ * as defined by the BatchConfig.
+ */
+class BatchPoller(
+    private val serverUrl: HttpUrl,
+    private val httpCallFactory: Call.Factory,
+    private val scalarTypeAdapters: ScalarTypeAdapters,
+    private val logger: ApolloLogger,
+    private val batchConfig: BatchConfig,
+    private val dispatcher: Executor,
+    private val scheduledExecutorService: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
```

Comment:
Did you consider reusing the same `dispatcher` that is used above to save a thread? Mainly curious about the decision to go with a separate thread for something that doesn't do a lot of work?

Reply:
Since that dispatcher is passed by the application, I didn't want to make assumptions on it. Thought it was safer to control this single thread that needs to run for a long period of time internally, and make sure it doesn't reduce the number of threads available in the application owned dispatcher.

Reply:
Makes sense 👍

Reply:
I would have made this an implementation details instead of a constructor property since it's always a `newSingleThreadScheduledExecutor` ultimately but I'll leave that decision to you.

Reply:
I was anticipating the unit tests here :D but will see if I can hide it